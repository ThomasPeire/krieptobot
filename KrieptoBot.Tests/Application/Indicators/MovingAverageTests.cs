using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using KrieptoBot.Application.Indicators;
using KrieptoBot.DataVisualizer;
using KrieptoBot.DataVisualizer.Extensions;
using KrieptoBot.Domain.Trading.ValueObjects;
using KrieptoBot.Infrastructure.Bitvavo.Dtos;
using KrieptoBot.Infrastructure.Bitvavo.Extensions;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using Plotly.NET;
using Snapshooter.NUnit;
using Color = System.Drawing.Color;

namespace KrieptoBot.Tests.Application.Indicators;

internal class MovingAverageTests
{
    private IEnumerable<Candle> _candles;

    [SetUp]
    public void Setup()
    {
        InitCandles();
    }

    private void InitCandles()
    {
        var candlesJson = File.ReadAllText(@"./MockData/Bitvavo/candles_btc-eur.json");
        var deserializedCandles = JsonConvert.DeserializeObject(candlesJson) as JArray;
        _candles = deserializedCandles.Select(x =>
            new CandleDto
            {
                TimeStamp = x.Value<long>(0),
                Open = x.Value<decimal>(1),
                High = x.Value<decimal>(2),
                Low = x.Value<decimal>(3),
                Close = x.Value<decimal>(4),
                Volume = x.Value<decimal>(5)
            }.ConvertToKrieptoBotModel()).DistinctBy(x => x.TimeStamp);
    }

    [Test]
    public void MovingAverage_Should_CalculateMovingAverage()
    {
        var datetimeFrom = new DateTime(2021, 11, 01);
        var dateTimeTo = datetimeFrom.AddHours(8);

        var candlesToWorkWith = _candles
            .Where(x => x.TimeStamp >= datetimeFrom && x.TimeStamp <= dateTimeTo)
            .OrderBy(x => x.TimeStamp);

        var values5 = new MovingAverage().Calculate(candlesToWorkWith, 5);
        var values10 = new MovingAverage().Calculate(candlesToWorkWith, 14);

        values5 = candlesToWorkWith.ToDictionary(x => x.TimeStamp,
            x => values5.TryGetValue(x.TimeStamp, out var value) ? value : 0);

        Snapshot.Match(values5);
#if DEBUG
        var candleVisualizer = new CandlesVisualizer();
        var candleChart = candleVisualizer.Visualize(candlesToWorkWith);

        candleChart = candleChart.AddLineChart(values5, Color.Yellow);
        candleChart = candleChart.AddLineChart(values10, Color.Blue);
        candleChart.WithSize(1920, 1080).WithConfig(Config.init(Responsive: true)).Show();
#endif
    }
}