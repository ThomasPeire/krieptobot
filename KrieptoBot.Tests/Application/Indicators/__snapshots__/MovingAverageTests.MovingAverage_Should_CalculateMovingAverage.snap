{
  "2021-11-01T00:00:00Z": 53236.0,
  "2021-11-01T00:05:00Z": 53236.0,
  "2021-11-01T00:10:00Z": 53236.0,
  "2021-11-01T00:15:00Z": 53236.0,
  "2021-11-01T00:20:00Z": 53236.0,
  "2021-11-01T00:25:00Z": 53280.4,
  "2021-11-01T00:30:00Z": 53296.0,
  "2021-11-01T00:35:00Z": 53299.8,
  "2021-11-01T00:40:00Z": 53267.2,
  "2021-11-01T00:45:00Z": 53262.4,
  "2021-11-01T00:50:00Z": 53277.6,
  "2021-11-01T00:55:00Z": 53296.4,
  "2021-11-01T01:00:00Z": 53307.6,
  "2021-11-01T01:05:00Z": 53343.6,
  "2021-11-01T01:10:00Z": 53390.6,
  "2021-11-01T01:15:00Z": 53402.2,
  "2021-11-01T01:20:00Z": 53415.0,
  "2021-11-01T01:25:00Z": 53422.8,
  "2021-11-01T01:30:00Z": 53396.6,
  "2021-11-01T01:35:00Z": 53361.0,
  "2021-11-01T01:40:00Z": 53315.6,
  "2021-11-01T01:45:00Z": 53274.2,
  "2021-11-01T01:50:00Z": 53220.6,
  "2021-11-01T01:55:00Z": 53187.8,
  "2021-11-01T02:00:00Z": 53152.0,
  "2021-11-01T02:05:00Z": 53158.2,
  "2021-11-01T02:10:00Z": 53121.4,
  "2021-11-01T02:15:00Z": 52978.0,
  "2021-11-01T02:20:00Z": 52787.6,
  "2021-11-01T02:25:00Z": 52590.4,
  "2021-11-01T02:30:00Z": 52398.6,
  "2021-11-01T02:35:00Z": 52265.2,
  "2021-11-01T02:40:00Z": 52232.0,
  "2021-11-01T02:45:00Z": 52256.0,
  "2021-11-01T02:50:00Z": 52311.4,
  "2021-11-01T02:55:00Z": 52362.4,
  "2021-11-01T03:00:00Z": 52365.2,
  "2021-11-01T03:05:00Z": 52379.2,
  "2021-11-01T03:10:00Z": 52378.0,
  "2021-11-01T03:15:00Z": 52345.6,
  "2021-11-01T03:20:00Z": 52292.0,
  "2021-11-01T03:25:00Z": 52279.4,
  "2021-11-01T03:30:00Z": 52253.0,
  "2021-11-01T03:35:00Z": 52240.6,
  "2021-11-01T03:40:00Z": 52240.2,
  "2021-11-01T03:45:00Z": 52230.2,
  "2021-11-01T03:50:00Z": 52200.2,
  "2021-11-01T03:55:00Z": 52140.8,
  "2021-11-01T04:00:00Z": 52069.8,
  "2021-11-01T04:05:00Z": 51996.6,
  "2021-11-01T04:10:00Z": 51996.4,
  "2021-11-01T04:15:00Z": 51993.0,
  "2021-11-01T04:20:00Z": 52068.6,
  "2021-11-01T04:25:00Z": 52143.4,
  "2021-11-01T04:30:00Z": 52277.8,
  "2021-11-01T04:35:00Z": 52341.6,
  "2021-11-01T04:40:00Z": 52424.6,
  "2021-11-01T04:45:00Z": 52461.6,
  "2021-11-01T04:50:00Z": 52519.0,
  "2021-11-01T04:55:00Z": 52512.2,
  "2021-11-01T05:00:00Z": 52501.0,
  "2021-11-01T05:05:00Z": 52493.0,
  "2021-11-01T05:10:00Z": 52512.8,
  "2021-11-01T05:15:00Z": 52519.2,
  "2021-11-01T05:20:00Z": 52552.6,
  "2021-11-01T05:25:00Z": 52578.6,
  "2021-11-01T05:30:00Z": 52586.0,
  "2021-11-01T05:35:00Z": 52574.6,
  "2021-11-01T05:40:00Z": 52585.4,
  "2021-11-01T05:45:00Z": 52556.8,
  "2021-11-01T05:50:00Z": 52535.2,
  "2021-11-01T05:55:00Z": 52513.4,
  "2021-11-01T06:00:00Z": 52481.0,
  "2021-11-01T06:05:00Z": 52420.2,
  "2021-11-01T06:10:00Z": 52377.0,
  "2021-11-01T06:15:00Z": 52332.4,
  "2021-11-01T06:20:00Z": 52313.6,
  "2021-11-01T06:25:00Z": 52343.6,
  "2021-11-01T06:30:00Z": 52394.8,
  "2021-11-01T06:35:00Z": 52487.2,
  "2021-11-01T06:40:00Z": 52582.6,
  "2021-11-01T06:45:00Z": 52627.0,
  "2021-11-01T06:50:00Z": 52641.6,
  "2021-11-01T06:55:00Z": 52653.4,
  "2021-11-01T07:00:00Z": 52723.4,
  "2021-11-01T07:05:00Z": 52838.0,
  "2021-11-01T07:10:00Z": 53082.4,
  "2021-11-01T07:15:00Z": 53310.4,
  "2021-11-01T07:20:00Z": 53553.8,
  "2021-11-01T07:25:00Z": 53676.2,
  "2021-11-01T07:30:00Z": 53777.8,
  "2021-11-01T07:35:00Z": 53746.8,
  "2021-11-01T07:40:00Z": 53740.2,
  "2021-11-01T07:45:00Z": 53727.8,
  "2021-11-01T07:50:00Z": 53746.0,
  "2021-11-01T07:55:00Z": 53743.4,
  "2021-11-01T08:00:00Z": 53797.6
}
